Add a function to display elapsed recording time on a label

gort818 opened the pull request:
Add function that starts counting the time you are recording and displays the time in a label on the headerbar.

gort818 commented:
This will need some changes with the formatting of the label any help appreciated.. trying to center the label on the headbar and having some difficulties.

gort818 commented on src/ui/application-window.vala:
```diff
@@ -251,6 +254,22 @@ namespace Peek.Ui {
+      double seconds;
+      Timer timer = new Timer ();
+      while (is_recording && stop_button.sensitive == true) {
+        seconds=(int)timer.elapsed();
```
Old habits die hard :(

gort818 commented on src/ui/application-window.vala:
```diff
@@ -251,6 +254,22 @@ namespace Peek.Ui {
+    
+    async void update_time () {
+      double seconds;
+      Timer timer = new Timer ();
```
What about using add_seconds_full? https://valadoc.org/glib-2.0/GLib.Timeout.add_seconds_full.html

gort818 replied:
I initalially choose  not to use Timeout because
>Note that timeout functions may be delayed, due to the processing of other event sources. Thus they should not be relied on for precise timing

But I will give a test!

phw replied:
That's true. You could still store the start time (in seconds) and on each interval calculate the elapsed time. That way a small time slip in the Timeout calls does not lead to wrongly displayed time.

phw replied:
And yes, add_seconds_full could be tried. There is some info that the first call can be delayed by up to one second, though. Never tried it.

gort818 commented:
Ok I gave it try everything seems to be working, let me know what you think.. now how to make it look pretty??

phw commented on src/ui/application-window.vala:
```diff
@@ -251,6 +254,23 @@ namespace Peek.Ui {
     }
+    
+    private void update_time () {
+      int seconds=0;
```
Old habbits :p Also two lines lower

gort818 replied:
Oh yeah It is automatic.. like autopilot :)

phw replied:
Please update the spacing around `=` and `+=` below, than this PR looks great code wise.

phw commented on src/ui/application-window.vala:
```diff
@@ -251,6 +254,23 @@ namespace Peek.Ui {
+      int seconds=0;
+      Timeout.add_seconds_full (GLib.Priority.LOW,1, () => {
+        seconds +=1;
+        if (is_recording && stop_button.sensitive == true) {
```
I am just curious, why is stop_button.sensitive needed to be checked?

gort818 replied:
To be honest It is not needed, I was doing some testing and left it in there. It will be removed.

phw commented on src/ui/application-window.vala:
```diff
@@ -251,6 +254,23 @@ namespace Peek.Ui {
+    
+    private void update_time () {
+      int seconds=0;
+      Timeout.add_seconds_full (GLib.Priority.LOW,1, () => {
```
So let's see how this works out. If we find out the time will vary too much for longer recordings, we can first maybe try to increase the priority here or do the time calculation on each call based on a start time.

phw commented on ui/application-window.ui:
```diff
@@ -198,6 +198,17 @@ Author: Philipp Wolfer <ph.wolfer@gmail.com>
+            <child>
+              <object class="GtkLabel" id="time_label">
+                <property name="can_focus">False</property>
+                <property name="label" translatable="yes">label</property>
```
You should mark this as non-translatable. Also please change the default "label" into "00:00", that way it looks nicer in glade.

phw commented on src/ui/application-window.vala:
```diff
@@ -369,6 +389,7 @@ namespace Peek.Ui {
             hide_handler = delay_indicator.hide.connect (() => {
               delay_indicator.disconnect (hide_handler);
               start_recording ();
+              update_time ();
```
I would suggest to call this from inside `start_recording ()`. Also note that  `start_recording ()` gets called on two locations, currently your branch works only if one has configured displaying the start recording timeout.

gort818 commented:
Ok now to get started on the ui changes.. I will have some later for you to take a look at.

gort818 commented:
Here are the buttons left aligned

![screenshot from 2017-11-17 17-33-06](https://user-images.githubusercontent.com/9098121/32975573-72488a7a-cbbd-11e7-8144-91226ba511ce.png)
![screenshot from 2017-11-17 17-32-48](https://user-images.githubusercontent.com/9098121/32975574-72600c72-cbbd-11e7-8154-a5039d9ecce6.png)

gort818 commented:
![screenshot from 2017-11-17 19-02-07](https://user-images.githubusercontent.com/9098121/32976399-dee37b8e-cbc9-11e7-8631-3ef3b66fd681.png)
![screenshot from 2017-11-17 19-01-45](https://user-images.githubusercontent.com/9098121/32976400-df13203c-cbc9-11e7-9cd8-5100670a706f.png)

Ok I'm getting somewhere. I changed the headerbar (removed custom title, added the buttonbox and all the goodies) and used the headerbar title as the time display as it is always centered. 

Is this something that is feasible it all seems to work, but you might have a reason for using the custom title??

I will put these changes in a different branch, as I have not merged them yet. But let me know if this can work.

gort818 commented:
![screenshot from 2017-11-17 19-17-02](https://user-images.githubusercontent.com/9098121/32976500-0179205c-cbcc-11e7-9d24-6f4a6ee03f1f.png)
![screenshot from 2017-11-17 19-16-40](https://user-images.githubusercontent.com/9098121/32976501-018fa99e-cbcc-11e7-8d82-41816a7b8cba.png)
Here are some shots with Gnome so we can see how it looks with the application menu.

phw commented:
Good idea! Yes, using the app title makes sense. I think the only reason I disabled it was that it made problems with the centered button layout. But since we don't use that anymore let's use the title for the time.

This also makes it easy to display two lines of information should we need it.

gort818 commented:
Ok awesome I updated the branch take a took.. I think it is looking really great!

phw commented:
Great work, tried it and it is really cool. I have to get used to the new button location, but the recording time is really useful I must admit :) Also it is good to have it centered, it is easy to spot then.

Now please just fix the spacing around `=` and `+=` (lines 257 and 259) and this is good to get merged :)

FYI: I want to get this merged and see how my fixes for https://github.com/phw/peek/issues/199 work out. Then give it a week to see if any issues pop up and give some more time to get translations updated. And next weekend I want to finally release 1.2.0 :)

gort818 commented:
Ok done, very nice :)!

phw commented:
Thanks a lot again, merged
